Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h

```diff
@@ -8,127 +8,329 @@
 
 #pragma once
 
-#include <type_traits>
 #include <AzCore/base.h>
-#include <AzCore/RTTI/TypeInfoSimple.h>
-#include <AzCore/std/limits.h>
+#include <AzCore/Outcome/Outcome.h>
+#include <AzCore/Platform.h>
+#include <AzCore/RTTI/RTTI.h>
+#include <AzCore/std/chrono/chrono.h>
+#include <AzCore/std/containers/span.h>
+#include <AzCore/std/containers/variant.h>
 #include <AzCore/std/string/string_view.h>
-#include <AzCore/Casting/numeric_cast.h>
+#include <AzCore/std/utils.h>
+
+namespace AZ::IO
+{
+    class GenericStream;
+}
 
 namespace AZ::Metrics
 {
-    //! Simple hash structure based on DJB2a to generate event IDs at compile time
-    class EventNameHash
+    constexpr AZStd::string_view ArgsKey = "args";
+
+    //! Event fields structure that can be used to record event argument data
+    //! NOTE: the EventValue does not own any data
+    //! Any string or array data must outlive the EventValue
+    struct EventValue;
+
+    //! Encapsulates an EventField paired with a field name.
+    //! Represents a single json key, value pair field
+    struct EventField;
+
+    //! std::span requires a complete type, so an EventValue pointer and a size member
+    //! is stored in the public API exposes a span interface
+    //! The reason why EventValue can't be complete is because it is a recursive type
+    //! that has a reference to an array of child event values
+    struct EventArray
     {
-    public:
-        constexpr explicit EventNameHash(AZStd::string_view name)
-        {
-            for (auto c : name)
-            {
-                m_hash = static_cast<uint32_t>(((static_cast<uint64_t>(m_hash) << 5) + m_hash) ^ c);
-            }
-        }
-
-        constexpr bool operator==(const EventNameHash& rhs) const
-        {
-            return m_hash == rhs.m_hash;
-        }
-
-        constexpr bool operator!=(const EventNameHash& rhs) const
-        {
-            return m_hash != rhs.m_hash;
-        }
+        constexpr EventArray() = default;
+        constexpr explicit EventArray(AZStd::span<EventValue> arrayValues);
+
+        constexpr void SetArrayValues (AZStd::span<EventValue> arrayValues);
+        constexpr AZStd::span<EventValue> GetArrayValues() const;
+
+    private:
+        EventValue* m_arrayAddress{};
+        size_t m_arraySize{};
+    };
+
+    //! std::span requires a complete type, so an EventField pointer and a size member
```

Comment:
same comment about AZStd::span here

Reply:
I'll mention both AZStd::span and std::span here